packages/sdk/schema/src/testing/generator.ts

```diff
@@ -0,0 +1,162 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { Effect, pipe } from 'effect';
+
+import { type EchoDatabase } from '@dxos/echo-db';
+import {
+  create,
+  type S,
+  type ReactiveObject,
+  type ExcludeId,
+  type BaseObject,
+  FormatEnum,
+  GeneratorAnnotationId,
+  type JsonSchemaType,
+  getSchemaReference,
+  getTypename,
+} from '@dxos/echo-schema';
+import { AST, findAnnotation } from '@dxos/effect';
+import { invariant } from '@dxos/invariant';
+import { log } from '@dxos/log';
+import { faker } from '@dxos/random';
+import { getDeep } from '@dxos/util';
+
+import { getSchemaProperties } from '../properties';
+
+// TODO(burdon): AUDIT (@dmytro)
+//  - New TypedObject syntax.
+//    - Define low-level types in S.Struct (e.g., in react-table).
+//  - Audit all low-level types (e.g., ReactiveObject, AbstractTypedObject, MutableObject: see echo-schema/docs).
+//  - Implement basic "comment required" TODOs.
+
+// TODO(burdon): Agent pipeline (@dmytro)
+//  - Generators: https://effect.website/docs/getting-started/using-generators/
+
+// TODO(burdon): Replace echo-generator.
+// TODO(burdon): Delete core/agent, experimental/agent-functions.
+// TODO(burdon): Generate views, tables, and SCRIPTS/FUNCTIONS.
+// TODO(burdon): Generate test documents, sketches, sheets.
+
+/**
+ * Set properties based on generator annotation.
+ */
+export const setProps = <T extends BaseObject>(type: S.Schema<T>) => {
+  return (initial: ExcludeId<T> = {} as ExcludeId<T>): Effect.Effect<ExcludeId<T>> => {
+    const obj = getSchemaProperties<T>(type.ast).reduce<ExcludeId<T>>((data, property) => {
+      if (data[property.name] === undefined) {
+        const gen = findAnnotation<string>(property.ast, GeneratorAnnotationId);
+        if (gen) {
+          const fn = getDeep<() => any>({ faker }, gen.split('.'));
+          if (fn) {
+            data[property.name] = fn() as any;
+          }
+        } else if (!property.optional) {
+          log.warn('missing generator for required property', { property });
+        }
+      }
+
+      return data;
+    }, initial);
+
+    return Effect.succeed(obj);
+  };
+};
+
+/**
+ * Set references.
+ */
+export const setReferences = <T extends BaseObject>(type: S.Schema<T>, db: EchoDatabase) => {
+  return (obj: BaseObject<T>) => {
+    return Effect.promise(async () => {
+      for (const property of getSchemaProperties<T>(type.ast)) {
+        if (property.format === FormatEnum.Ref) {
+          const jsonSchema = findAnnotation<JsonSchemaType>(property.ast, AST.JSONSchemaAnnotationId);
```

Comment:
I don't think we should be accessing `JSONSchemaAnnotationId` anywhere but the serializer.

I would just convert the effect schema to JSON-schema using the standard serializer and work with that.